Rethink multiline annotations

cage2 opened the pull request:
See : #59

Commit:
- starting with a new method to render multiline annotations

  each annotation (the overlay, actually) now has a property 'position
  and its  value indicates which  positions the annotations holds  in a
  "chain"  of  annotations.

  Even  if rendered  separately  each chain  (better called  "group"?)
  represent a single annotation.

  The last annotation in the chain has position's value equal to -1.

  If  the set  of a  group/chain  is formed  by only  one element  the
  position's value is -1 as well.

Commit:
Merge branch 'master' into rethink-multiline-annotations

Commit:
- fixed 'annotate-previous-annotation-ends' and 'annotate-next-annotation-ends'
  we was checking for an annotation but 'annotate-annotation-at' returned a list
  instead, now the latter return an annotation or nil.

Commit:
Merge branch 'master' into rethink-multiline-annotations

Commit:
- saving annotations with newlines seems to works;
- fixed 'annotate-move-next-annotation' and 'annotate-move-previous-annotation';
- fixed 'annotate-previous-annotation-end' a variable incremented then was
  added an harmful on to it.

Commit:
Merge branch 'master' into rethink-multiline-annotations

Commit:
- fixed exporting and integrations of annotations;
- prevent a newline to be annotated (at creation time).

bastibe commented on annotate.el:
```diff
@@ -69,9 +72,9 @@ See https://github.com/bastibe/annotate.el/ for documentation."
 (define-key annotate-mode-map (kbd "C-c C-s") 'annotate-show-annotation-summary)
 
-(define-key annotate-mode-map (kbd "C-c ]") 'annotate-next-annotation)
+(define-key annotate-mode-map (kbd "C-c ]") 'annotate-move-next-annotation)
```
I think "move" is ambiguous, as it might be interpreted as moving the annotation. Perhaps "goto" is better?
```suggestion
(define-key annotate-mode-map (kbd "C-c ]") 'annotate-goto-next-annotation)
```

cage2 replied:
Hello!

I agree it is better to rename, this is a case where 'goto' is not considered harmful! ;-D

bastibe commented on annotate.el:
```diff
@@ -69,9 +72,9 @@ See https://github.com/bastibe/annotate.el/ for documentation."
+(define-key annotate-mode-map (kbd "C-c ]") 'annotate-move-next-annotation)
 
-(define-key annotate-mode-map (kbd "C-c [") 'annotate-previous-annotation)
+(define-key annotate-mode-map (kbd "C-c [") 'annotate-move-previous-annotation)
```
See above
```suggestion
(define-key annotate-mode-map (kbd "C-c [") 'annotate-goto-previous-annotation)
```

bastibe commented on annotate.el:
```diff
@@ -367,7 +376,7 @@ modified (for example a newline is inserted)."
       (set-buffer-modified-p t))))
 
-(defun annotate-next-annotation ()
+(defun annotate-move-next-annotation ()
```
```suggestion
(defun annotate-goto-next-annotation ()
```

bastibe commented on annotate.el:
```diff
@@ -387,7 +396,7 @@ modified (for example a newline is inserted)."
       (goto-char (overlay-start (nth 0 overlays))))))
 
-(defun annotate-previous-annotation ()
+(defun annotate-move-previous-annotation ()
```
```suggestion
(defun annotate-goto-previous-annotation ()
```

bastibe reviewed (commented):
Wow, another big change!

In general, this looks good. What I don't understand yet: What annotation-prop-chain-position is assigned to single-line annotations?

In other words, are there now two "classes" of annotations, one a chain and one a single annotation, that are handled differently by the code, or is a single-line annotation simply a chain of length 1?

Otherwise, I applaud your efforts, but haven't been able to follow all parts of the code. A few comments would probably help me a lot!

bastibe commented:
Also, it occurred to me that it might be possible to manually set the `underline` face of annotations to `nil` instead of having to split the overlays. Or possibly, add a single-character overlay with a `nil` underline on the newline. Did you check if something like that is possible?

Commit:
- optimized 'annotate-previous-annotation-ends' and 'annotate-next-annotation-starts'.

Commit:
- changed function names:
  'annotate-move-previous-annotation' -> 'annotate-goto-previous-annotation';
  'annotate-move-next-annotation'     -> 'annotate-goto-next-annotation'

cage2 commented:
Hello Bastian!

> Also, it occurred to me that it might be possible to manually set the `underline` face of annotations to `nil` instead of having to split the overlays. Or possibly, add a single-character overlay with a `nil` underline on the newline. Did you check if something like that is possible?

I tried the first with no luck but the second seems a good point, let me explore this idea! 

Bye!
C.

cage2 commented:
Hi!

> Wow, another big change!

Thank you! :)
 
> In general, this looks good. What I don't understand yet: What annotation-prop-chain-position is assigned to single-line annotations?
> 
> In other words, are there now two "classes" of annotations, one a chain and one a single annotation, that are handled differently by the code, or is a single-line annotation simply a chain of length 1?

the second, i assigned the position `-1` to a chain made of single annotation, therefore the member of this this particular class of "chains" is both the first and the last. Seems the right thing to do but i fear this was quite arbitrary in retrospective, what do you think?

> 
> Otherwise, I applaud your efforts, but haven't been able to follow all parts of the code. A few comments would probably help me a lot!

Sure, let me add some more comments!

Bye!
C.

bastibe commented:
Cool! That makes sense, then.

For some reason, the numbering and specifically the `-1` feels off to me. But I can't quite put my finger on the reason, or a better solution. Maybe `nil` or a dedicated symbol might be better than `-1`? But then that's just playing games with syntax, and not really meaningful. I wonder if there is a different way, perhaps by linking the overlays explicitly with a linked list, where each overlay refers to the next or previous overlay in a property or something like that.

Anyway, these are just idle thoughts, and not important unless you find the topic interesting as a coding puzzle.

cage2 commented:
On Tue, Mar 03, 2020 at 11:52:10PM -0800, Bastian Bechtold wrote:

Hello Bastian!

> Cool! That makes sense, then.

Well i  feel somehow is  a bit  "hacky" but at  least it does  fix the
problem and, with my surprise it the fact that an annotation (a single
overlay,  i mean)  can lay  on a  single line  made some  code simpler
(look, for example  at the diff generating code). And  this was totally
unintentional of course! ;-D ;-D

> For some reason,  the numbering and specifically the  `-1` feels off
> to me. But  I can't quite put  my finger on the reason,  or a better
> solution. Maybe  `nil` or  a dedicated symbol  might be  better than
> `-1`?

I do not like 'nil' because (if i remember correctly) if a property of
annotation does  not exists the expression  (overlay-get-property ...)
values  nil as  well,  moreover i  prefer  to keep  the  value of  the
properties of  the same type,  but these,  as you wrote,  are actually
matter of taste! :)

> But  then that's  just playing  games  with syntax,  and not  really
> meaningful. I wonder if there is a different way, perhaps by linking
> the  overlays explicitly  with  a linked  list,  where each  overlay
> refers to  the next or previous  overlay in a property  or something
> like that.

This  would be  a better  (or at  least cleaner)  solution, but  i was
scared to use buffer local variables and other thing i never used :)

> Anyway, these are  just idle thoughts, and not  important unless you
> find the topic interesting as a coding puzzle.

Probably  using an  external data  structure would  be interesting,  i
would   open   an   issue    just   as   remainder   mentioning   this
improvement. Maybe this could attract  to the project some other coder,
you never can tell! :)

Bye!
C.

bastibe commented:
True, I hadn't thought of that `nil` problem. And the linked list proposal feels like it could break easily.

>  i would   open   an   issue    just   as   remainder   mentioning   this improvement. Maybe this could attract  to the project some other coder, you never can tell! :)

Alright!

Commit:
Merge branch 'master' into rethink-multiline-annotations

Commit:
- fixed return value of 'maybe-force-newline-policy'.

cage2 commented:
Hello Bastian! 
I locally merged the master branch into this one, things seems to works so far but i would like to spend a few hours using the package on an actual code base (mine, so if i break things nobody will be upset, except me! ;-) )

Bye!
C.

Commit:
- updated documentation.

Commit:
- fixed removing of annotation in 'annotate-before-change-fn'

  sometimes  when the  annotated text  is changed  we could  remove an
  overlay  that is  part of  an annotation  chain. This  is ok  but if
  happens to delete  the last element of a chain  we should ensure the
  the preceding overlay takes is place, if such overlay exists.

cage2 commented:
Hi Bastian!

After last fix I think i am more or less ready to merge this :)

Bye!
C.

bastibe commented:
OK, cool!

Although we should probably increment the minor version number instead of the bugfix number, as these changes are more akin to new features than to bug fixes. (In fact, we should probably think about releasing a new major version soon, as there were a few backwards-incompatible changes.)

cage2 commented:
On Tue, Mar 10, 2020 at 12:41:02AM -0700, Bastian Bechtold wrote:

Hello Bastian!

> OK, cool!

> Although  we  should probably  increment  the  minor version  number
> instead of the bugfix number, as  these changes are more akin to new
> features than to bug fixes.

I totally agree, should we number this version 0.6.0?

> (In fact, we should probably think about
> releasing  a   new  major  version   soon,  as  there  were   a  few
> backwards-incompatible changes.)

Do you think is time for version 1.0?

Bye!
C.

bastibe commented:
> I totally agree, should we number this version 0.6.0?

Yes, I think that's a good idea!

> Do you think is time for version 1.0?

I'll leave that up to you, honestly. I'd wait with 1.0 until you're happy with the state of the library and have implemented all of your planned big changes. At any rate, I would let the project rest for a month or so, and see how the current state works in day-to-day use before changing the major version.

cage2 commented:
On Wed, Mar 11, 2020 at 02:51:08AM -0700, Bastian Bechtold wrote:

Hello Bastian!

> > I totally agree, should we number this version 0.6.0?
>
> Yes, I think that's a good idea!

Ok!

> > Do you think is time for version 1.0?
>
> I'll leave that up to you, honestly.

Wow, that is a big responsability! Thank  you for the trust, i hope it
will be deserved! :)

> I'd wait with  1.0 until you're happy with the  state of the library
> and have implemented all of your planned big changes. At any rate, I
> would  let the  project rest  for a  month or  so, and  see how  the
> current  state works  in day-to-day  use before  changing the  major
> version.

The package appears more or less  feature-complete to me now, but i am
quite sure  some bugs (hopefully a  small number) are going  to pop-up
soon,  given  that  the  changes  that this  branch  brings  are  very
pervasive.  So, basically i agree to pospone the 1.0 to the future for
at least a  month with no bugs reports. In  the meanwhile, let's merge
this branch and let the code runs!  :)

Bye!
C.

Commit:
- changed version number.
